src/orange/OrangeParams.cc

```diff
@@ -180,15 +178,63 @@ OrangeParams::OrangeParams(OrangeInput input)
             }
             volume_labels.push_back(std::move(volume_label));
         }
-
-        bbox_ = u.bbox;
     }
 
     surf_labels_ = LabelIdMultiMap<SurfaceId>{std::move(surface_labels)};
     vol_labels_ = LabelIdMultiMap<VolumeId>{std::move(volume_labels)};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Advanced usage: construct from explicit host data.
+ *
+ * Volume and surface labels must be unique for the time being.
+ */
+OrangeParams::OrangeParams(OrangeInput input)
+{
+    CELER_VALIDATE(!input.units.empty(), << "input geometry has no universes");
+
+    HostVal<OrangeParamsData> host_data;
 
-    supports_safety_ = host_data.simple_unit[SimpleUnitId{0}].simple_safety;
-    bbox_ = input.units.front().bbox;
+    // Copy data from OrangeInput
+    host_data.scalars.max_level = input.max_level;
+    make_builder(&host_data.universe_types)
+        .insert_back(input.universe_types.begin(), input.universe_types.end());
+    make_builder(&host_data.universe_indices)
+        .insert_back(input.universe_indices.begin(),
+                     input.universe_indices.end());
+
+    // Insert simple units
+    insert_simple_units(host_data, input);
```

Comment:
Use `this` pointers for non-static member functions for clarity.
```suggestion
    this->insert_simple_units(host_data, input);
```